fix: Timeout crash because of chunked_output len

bogdansalyp opened the pull request:
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses - closes #2554 

#### Changelog
What are the changes made in this PR?
* replaced `torch.chunk` with `torch.tensor_split`
#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

pbontrager reviewed (commented):
Why not just use torch.tensor_split? Also, we should test what happens in the case of 7 tokens, tensor split will return an 8th empty tensor.

bogdansalyp commented:
> Why not just use torch.tensor_split? Also, we should test what happens in the case of 7 tokens, tensor split will return an 8th empty tensor.

![image](https://github.com/user-attachments/assets/d90210ec-57f7-4d04-8d62-312158d310e0)

Wow you're right, I didn't know this func differs. Will remove my module from the PR and replace with this func
Regarding splitting 7 into 8 - it gives an empty tensor as the 8th element. Might work out of box, will check tomorrow

bogdansalyp commented:
@pbontrager @felipemello1 review pls

Added `tensor_split`, added tests for transformer `chunked_output` func (covered simple case, 49 len and 7 len into 8 chunks)

felipemello1 reviewed (commented):
LGTM, thank you! Just left one comment.

Is there any chance that you can run it with compile=True for just a few steps (~10) and share the run stats from wandb (or some other logger), to make sure things look fine?

bogdansalyp commented on torchtune/modules/loss/ce_chunked_output_loss.py:
```diff
@@ -57,7 +57,7 @@ def forward(self, logits: List[torch.Tensor], labels: torch.Tensor) -> torch.Ten
             >>>
             >>> h = torch.tensor([bsz, num_tokens, dim])
-            >>> output_chunks = [model.output(chunk) for chunk in h.chunk(num_chunks, dim=1)]
+            >>> output_chunks = [model.output(chunk) for chunk in h.tensor_split(hum_chunks, dim=1)]
```
fixed

bogdansalyp commented on tests/torchtune/modules/test_transformer_decoder.py:
```diff
@@ -369,6 +378,45 @@ def test_forward(
+        input_params: Tuple[int, int, int],
+        decoder: TransformerDecoder,
+    ) -> None:
+        """Checks chunked output corner case (torch.chunk would've split in 7)."""
```
renamed funcs and expanded docstring

bogdansalyp commented:
<img width="1613" alt="image" src="https://github.com/user-attachments/assets/2edece45-d046-4ed1-bdfb-9d6b0bd2c1f3" />
<img width="790" alt="image" src="https://github.com/user-attachments/assets/ed3e9235-bf9f-4b4d-b055-c06c57ce8b6f" />


launched with `compile=True`

bogdansalyp commented:
@felipemello1 addressed your change requests & attached wandb
